src/goto-instrument/cover.cpp

```diff
@@ -218,6 +218,299 @@ void collect_operands(const exprt &src, std::vector<exprt> &dest)
+    {
+      return result;
+    }
+    // the expansion of "==" is ">" and "<"
+    exprt e1(ID_gt);
+    e1.type().id(src.type().id());
+    e1.operands().push_back(src.op0());
+    e1.operands().push_back(src.op1());
+    result.insert(e1);
+
+    exprt e2(ID_lt);
+    e2.type().id(src.type().id());
+    e2.operands().push_back(src.op0());
+    e2.operands().push_back(src.op1());
+    result.insert(e2);
+  }
+
+  return result;
+}
+
+/*******************************************************************\
+
+Function: ordered_negation
+
+  Inputs:
+
+ Outputs:
+
+ Purpose:
+
+\*******************************************************************/
+
+std::set<exprt> ordered_negation(const exprt &src)
+{
+  std::set<exprt> result;
+  // the negation of "==" is "<" and ">"
+  if(src.id()==ID_equal)
+  {
+    exprt e1(src);
+    e1.id(ID_lt);
+    result.insert(e1);
+
+    exprt e2(src);
+    e2.id(ID_gt);
+    result.insert(e2);
+  }
+  // the negation of "<" is "==" and ">"
+  else if(src.id()==ID_lt)
+  {
+    exprt e1(src);
+    e1.id(ID_equal);
+    result.insert(e1);
+
+    exprt e2(src);
+    e2.id(ID_gt);
+    result.insert(e2);
+  }
+  // the negation of "<=" is ">"
+  else if(src.id()==ID_le)
+  {
+    exprt e1(src);
+    e1.id(ID_gt);
+    result.insert(e1);
+  }
+  // the negation of ">=" is "<"
+  else if(src.id()==ID_ge)
+  {
+    exprt e1(src);
+    e1.id(ID_lt);
+    result.insert(e1);
+  }
+  // the negation of ">" is "==" and "<"
+  else if(src.id()==ID_gt)
+  {
+    exprt e1(src);
+    e1.id(ID_equal);
+    result.insert(e1);
+
+    exprt e2(src);
+    e2.id(ID_lt);
+    result.insert(e2);
+  }
+  // the negation of "!=" is "=="
+  else if(src.id()==ID_notequal)
+  {
+    exprt e1(src);
+    e1.id(ID_equal);
+    result.insert(e1);
+  }
+
+  return result;
+}
+
+/*******************************************************************\
+
+Function: is_arithmetic_predicate
+
+  Inputs:
+
+ Outputs:
+
+ Purpose:
+
+\*******************************************************************/
+
+bool is_arithmetic_predicate(const exprt &src)
+{
+  return (src.id()==ID_lt
+          || src.id()==ID_le
+          || src.id()==ID_equal
+          || src.id()==ID_ge
+          || src.id()==ID_gt
+          || src.id()==ID_notequal);
+}
+
+/*******************************************************************\
+
+Function: replacement_and_conjunction
+
+  Inputs:
+
+ Outputs:
+
+ Purpose: To replace the i-th expr of ''operands'' with each
+          expr inside ''replacement_exprs''.
+
+\*******************************************************************/
+
+std::set<exprt> replacement_and_conjunction(
+  const std::set<exprt> &replacement_exprs,
+  const std::vector<exprt> &operands,
+  const std::size_t i)
+{
+  std::set<exprt> result;
+  for(auto &y : replacement_exprs)
+  {
+    std::vector<exprt> others;
+    for(std::size_t j=0; j<operands.size(); j++)
+      if(i!=j)
+        others.push_back(operands[j]);
+
+    others.push_back(y);
```

Comment:
Several problems here:
1. This assumes that the operands commute as the new element will be the last one.
2. What if operands.size()<=i?
3. Repeated push_back without prior reserve() is inefficient.

I would propose:
```
assert(operands.size()>i);
std::vector<exprt> others(operands);
others[i]=y;
```